### ISSUE TITLE
binance : The handshake operation timed out

### ISSUE THREAD

**Issue by nucwzq** [ccxt/ccxt#776]
Here is a new error....................

ccxt.base.errors.ExchangeNotAvailable: binance GET https://api.binance.com/api/v1/ticker/24hr?symbol=IOTABTC <urlopen error _ssl.c:492: The handshake operation timed out>

**Comment by kroitor** [comment]
Are you located in China?

**Comment by nucwzq** [comment]
@kroitor YES.......

**Comment by kroitor** [comment]
@nucwzq do you use windows?

**Comment by nucwzq** [comment]
@kroitor Yes, the win 10

**Comment by kroitor** [comment]
@nucwzq what's your version of Python?

**Comment by nucwzq** [comment]
@kroitor Python 2.7.6

**Comment by nucwzq** [comment]
@kroitor Whether I need to open VPN and to update the python version and the error can be addressed?

**Comment by kroitor** [comment]
@nucwzq most likely you need this: https://stackoverflow.com/a/31915123/5055465

**Comment by kroitor** [comment]
@nucwzq is this issue consistently repeating?

**Comment by nucwzq** [comment]
@kroitor yes, i just run it, it still have the error.

**Comment by nucwzq** [comment]
@kroitor Where to download the domain validation certificate as mentioned in : https://stackoverflow.com/a/31915123/5055465

**Comment by nucwzq** [comment]
@kroitor as mentioned  in : https://stackoverflow.com/a/31915123/5055465 
it uses its own located at ?\lib\site-packages\certifi\cacert.pem.
There is no certifi\cacert.pem in my python file path. 
![default](https://user-images.githubusercontent.com/21311417/33719132-bd5768b4-db9a-11e7-97d4-08ad2928123f.png)

**Comment by kroitor** [comment]
In fact... the timeout tells that they will drop your connection anyway... Does it work for you with a public CORS proxy?

```Python
exchange = ccxt.binance ({
    'apiKey': ...,
    'secret': ...,
    'proxy': 'https://cors-anywhere.herokuapp.com/',  #  ← last slash critical
    'enableRateLimit': True
})

# your code...
```

↑ Does it work?

**Comment by nucwzq** [comment]
@kroitor ccxt.base.errors.DDoSProtection: binance GET https://cors-anywhere.herokuapp.com/https://api.binance.com/api/v1/ticker/24hr?symbol=BNTBTC 429 Too Many Requests  The origin "*" has sent too many requests.
The number of requests is limited to 600 per 60 minutes. Please self-host CORS Anywhere if you need more quota. See https://github.com/Rob--W/cors-anywhere#demo-server

**Comment by nucwzq** [comment]
I run the above-mentioned code, there is the error:
ccxt.base.errors.DDoSProtection: binance GET https://cors-anywhere.herokuapp.com/https://api.binance.com/api/v1/ticker/24hr?symbol=BNTBTC 429 Too Many Requests  The origin "*" has sent too many requests.
The number of requests is limited to 600 per 60 minutes. Please self-host CORS Anywhere if you need more quota. See https://github.com/Rob--W/cors-anywhere#demo-server

**Comment by nucwzq** [comment]
I rerun the code with the following codes, and it is working now. However, i don't know whether it will occur the same error later.
exchange = ccxt.binance({
    'apiKey': '.....',
    'secret': '.....',
    #'proxy': 'https://cors-anywhere.herokuapp.com/',  #  ← last slash critical
    'enableRateLimit': True,
})

**Comment by nucwzq** [comment]
@kroitor I found the code sometimes is ok, but sometimes have the error:

ccxt.base.errors.ExchangeNotAvailable: binance GET https://api.binance.com/api/v1/ticker/24hr?symbol=BTSBTC <urlopen error _ssl.c:492: The handshake operation timed out>

**Comment by kroitor** [comment]
@nucwzq yep, we're observing the same occasionally... Still investigating into this.

**Comment by nucwzq** [comment]
@kroitor Ooook, I am waiting for your good news :)

**Comment by kroitor** [comment]
@nucwzq can you paste a full report, like here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue ?
When pasting code please use triple backtiks (&#96;&#96;&#96;) to enclose your code for readability like so:

&#96;&#96;&#96;
your code
&#96;&#96;&#96;

&#96;&#96;&#96;
verbose output
&#96;&#96;&#96;

**Comment by nucwzq** [comment]
Dear @kroitor i run it again, and the error hasn't appear now. If the error appear again, i will as soon as possible and as  detail as possible to describe the error report. Thank u.

**Comment by nucwzq** [comment]
@kroitor It appears the errors:

Traceback (most recent call last):
  File "F:/py_project/Crypto_data_analysis/test_first_qt.py", line 57, in <module>
    dcoinlist = crypto.get_dcoinlist(coinlist, droprate, volume, curr_price)##返回当天下降droprate的币
  File "F:\py_project\Crypto_data_analysis\crypto.py", line 32, in get_dcoinlist
    last_price = exchange.fetch_ticker(coin)['last']###访问公有API，获得当前最新要价
  File "C:\Python27\lib\site-packages\ccxt\binance.py", line 319, in fetch_ticker
    }, params))
  File "C:\Python27\lib\site-packages\ccxt\binance.py", line 615, in request
    response = self.fetch2(path, api, method, params, headers, body)
  File "C:\Python27\lib\site-packages\ccxt\base\exchange.py", line 274, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "C:\Python27\lib\site-packages\ccxt\base\exchange.py", line 335, in fetch
    self.raise_error(ExchangeNotAvailable, url, method, e)
  File "C:\Python27\lib\site-packages\ccxt\base\exchange.py", line 256, in raise_error
    details,
ccxt.base.errors.ExchangeNotAvailable: binance GET https://api.binance.com/api/v1/ticker/24hr?symbol=OAXBTC <urlopen error timed out> 


Where the code of the function "get_dcoinlist" is:

def get_dcoinlist(coinlist, drop_rate, volume, curr_price):
    dorp_coinlist = []
    for coin in coinlist:
        drop = exchange.fetch_ticker(coin)['change']
        btc_volume = exchange.fetch_ticker(coin)['quoteVolume']
        last_price = exchange.fetch_ticker(coin)['last']  ###This is the error line
        if drop < drop_rate and btc_volume > volume and last_price > curr_price :
            dorp_coinlist.append(coin)
    return dorp_coinlist

**Comment by kroitor** [comment]
```Python
def get_dcoinlist(coinlist, drop_rate, volume, curr_price):
    drop_coinlist = []
    for coin in coinlist:
        ticker = exchange.fetch_ticker(coin)  # ← don't duplicate it if you need more than one field
        drop = ticker['change']  # once you fetch it, you can access any field of a ticker
        btc_volume = ticker['quoteVolume']  # no need to fetch multiple times
        last_price = ticker['last'] 
        if drop < drop_rate and btc_volume > volume and last_price > curr_price :
            drop_coinlist.append(coin)
    return drop_coinlist
```

1. **Start paying attention to EVERY detail**. It means, literally, each letter and each symbol.
2. Read the Manual.
3. Browse through examples.
4. Learn to handle errors and exceptions in code with try/raise/except.
5. Learn and practice more Python, there's a lot of that on the internet. I can't physically teach everyone to code in Python even if I wanted to, sorry. Your question is like 2+2=5, you just have to learn Python first, and then do algorithmic trading with Python second.
6. In your code there is a DROP ←→ DORP typo.
7. I asked you to use backticks, you ignored that.
8. I asked you to read the guide on pasting **verbose** output, you ignored that.
9. I asked you to read the Manual, you ignored that too (because, if you did, you would not have the above question).
10. To me it looks like you're skipping words when you read my answers ) If I give you a link, I really believe you will read it. If you see other links there, you should read them too ) You can't cheat here, I'm afraid )

Basically, you're getting the same inaccessibility error from Binance as you did before, but you're not handling it properly. You have to learn to work with exceptions in Python if you want to be able to make money. Otherwise the program will not work in real life. You also have to learn how to debug code.

https://github.com/ccxt/ccxt/wiki/Manual#troubleshooting

**Comment by nucwzq** [comment]
@kroitor ..........Thank u!!!!!!
